whylog/constraints/__init__.py

```diff
@@ -1,43 +1,200 @@
-    def __init__(self, intervals):
-        pass
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
 
+    @classmethod
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+        For Front to display param names to user and then ask user for param contents.
+        """
+        return cls.PARAMS
 
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
 
+    def verify(cls, group_contents, param_dict):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
+
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
         pass
 
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+class TimeConstraint(AbstractConstraint):
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = sorted([MIN_DELTA, MAX_DELTA])
+
+    def __init__(self, groups, param_dict):
         """
-        Sets Minimum and maximum difference between values of params (if values are numbers).
+        :param groups: First element of groups is a group with earlier date.
+
+        I.e:
+        TimeConstraint(
+            [(0, 1), (2, 1)]
+            {'min_delta': datetime.timedelta(seconds=1), 'max_delta': datetime.timedelta(seconds=10)},
+        )
         """
-        pass
+        super(TimeConstraint, self).__init__(groups, param_dict)
+
+    def _check_constructor_params(self):
+        param_names = set(self.params.keys())
+        correct_param_names = set(self.get_param_names())
+        if (self.MIN_DELTA not in param_names and self.MAX_DELTA not in param_names)\
+                or param_names - correct_param_names:
+            raise ConstructorParamsError(self.TYPE, correct_param_names, param_names)
+
+    @classmethod
+    def verify(cls, group_contents, param_dict):
+        #TODO
```

Comment:
implement it and test it. Adding a class that doesn't work because 3 or 4 lines are missing is detrimental.

Actually I strongly suspect that current implementation will have poor performance due to the fact that this class has not been fully designed before implementation and now that we reviewed it and fixed it, it's going to go to trash.

This is one of two most important performance bottlenecks of the entire program. It must work very well and you cannot just delay half of its implementation for later.